Feature/mouse and keyboard

Zargess opened the pull request:
Made mouse and keyboard event based. Made enums for each key on mouse and keyboard.

Commit:
Mouse btn click and release

Commit:
Mouse scrolling

Commit:
Keyboard click and release

miniwolf commented on Engine/src/mini/input/Keyboard.java:
```diff
@@ -12,13 +14,24 @@
+        KeyboardKey k = getKey(key);
+
+        for (KeyboardListener listener : listeners) {
+            if (action == 0) {
```
Kunne man ikke bruge noget ACTION enums, i forhold til hvis man tænker på noget getKeyState, getMouseState(), eller måske giver det ikke mening?

miniwolf replied:
Det samme her som tidligere, action boolean uden for forløkken.

miniwolf commented on Engine/src/mini/input/Keyboard.java:
```diff
@@ -42,4 +71,13 @@ public static boolean isKeyUp(int key) {
     }
+
+    private static KeyboardKey getKey(int key) {
+        for(KeyboardKey k : KeyboardKey.values()) {
```
Formatting for (KeyboardKey k : KeyboardKey.values()), mellemrum mellem for :dagger:

miniwolf replied:
Her kunne man måske også bruge noget parallel lambda

miniwolf commented on Engine/src/mini/input/Mouse.java:
```diff
@@ -26,6 +29,16 @@
         actions.put(button, action);
+
+        MouseButton btn = MouseButton.values()[button];
+        for(MouseListener listener : listeners) {
```
Formatting on for

miniwolf commented on Engine/src/mini/input/Mouse.java:
```diff
@@ -47,6 +60,9 @@ public static void updatePosition(double xPos, double yPos) {
      */
     public static void updateScrollWheel(double xOffset, double yOffset) {
         wheelYOffset = yOffset;
+        for(MouseListener listener : listeners) {
```
Formatting on for

miniwolf commented on Game/src/main/MainApp.java:
```diff
@@ -20,14 +21,52 @@ public static void main(String[] args) {
+        */
         engine.renderEnvironmentMap(scene.getEnvironmentMap(), scene, new Vector3f(0, 2, 0));
 
+        Mouse.addMouseListener(new MouseListener() {
```
Jeg tænker at man nok hellere vil have en klasse med noget funktionalitet der implementerede MouseListener eller KeyboardListener. Kunne vi lave noget fancy med at enginen leder efter dem?
Sådan at vi ved at alle der implementere MouseListeners, som måske allerede er scene entities eller noget i den stil. Eller bliver det for magisk?

miniwolf commented on Game/src/main/MainApp.java:
```diff
@@ -3,6 +3,7 @@
 import loaders.LoaderSettings;
 import loaders.SceneLoader;
 import loaders.SceneLoaderFactory;
+import mini.input.*;
```
Nu er det her gjort i MainApp, så det betyder meget mindre, men at bruge * import er dårlig stil. Så for så vidt muligt, brug single imports.

miniwolf commented on Engine/src/mini/input/Mouse.java:
```diff
@@ -26,6 +29,16 @@
+            } else if (action == 0) {
+                listener.OnRelease(btn, lastXPos, lastYPos);
+            }
+        }
```
Pas her på med hvordan du optimere dine kald. Din forløkke kigger på action for hver eneste listener. Hvis du istedet havde ude omkring, ville det ikke kræve et if check hver gang. Måske er compileren god nok til at optimere det væk. But a small pointer

miniwolf reviewed (changes requested):
- Ændre formatting
- Gentænkt loop optimering, input er dyrt nok i forvejen.
- Og smid den idé med implements KeyboardListener på tegnebrættet. Men ikke nødvendig for et approve.

Commit:
Fixed feedback from pull request. Needs testing

miniwolf committed:
GLFW_REPEAT is added in the action enum.

Keyboard is now using KeyboardKey, Action pairs instead of the Integer
pair, to support the enum and not the value.

In Mouse the same thing is done.

The way we support this is by using a lookup of values on the enum.
@see DisplayManager for more references.

Proper camelCase has been applied in MouseListener.
